### ISSUE TITLE
Ensure packages are imported the same way (import "react" vs import("react"))

### ISSUE THREAD

**Issue by fregante** [import-js/eslint-plugin-import#2326]
`import()` is used as an indicator that the bundler can create an additional file (code splitting) instead of including the specified dependency in the main bundle.

Unfortunately it's possible that the same dependency is imported statically in another file, therefore either cancelling the effect or perhaps bundling it twice.

```js
// index
import './a.js'
import './b.js'
```

```js
// a.js
import { React } from 'react';
```

```js
// b.js
const { React } = await import('react');
```

### Rule request

Is it possible to ensure that a specific dependency is imported the same way across files automatically?

If not automatically, I wish there was a way to enforce a certain import method per-dependency, for example:

```js
"import/enforce-import-style": [
	"error",
	{
		static: ["react"],
		dynamic: ["my-little-heavy-dep"],
		never: ["never-import-me-directly"] // Probably already possible, unrelated
	}
]
```

**Comment by ljharb** [comment]
I suppose this is information that the ExportMap could gather - whether a module was required, statically imported, and/or dynamically imported - so there’s no technical barrier.

Bundling it twice wouldn’t happen; the bundler keeps track of that. Cancelling the effect is also something I believe isn’t a concern because you can indicate to the bundler whether a specific dependency should be always split, or never split, regardless of how it’s consumed.

In other words, I’m not sure this is something that needs handling, since the bundler itself should be able to do so.

**Comment by fregante** [comment]
> Bundling it twice wouldn’t happen; the bundler keeps track of that.

You're right, at least in [my simple test with webpack](https://github.com/fregante/webpack-sandbox/tree/main/treeshake-multiple-import-style).



> you can indicate to the bundler whether a specific dependency should be always split, or never split, regardless of how it’s consumed

How so? If that's actually possible, I think webpack would force the whole bundle to wait until every such modules are downloaded since `import 'module'` is set up to run synchronously. If it's a deep sub-dependency, then nothing can run until this sub-dependency has been loaded. Right?

Either way I'd rather prevent this from happening altogether rather than relying on the builder’s behavior.

If possible, automatic consistency would be great because I don't want another developer to undo my code-splitting effort by just importing `foo` statically.

**Comment by ljharb** [comment]
Either that, or the build process would error because you tried to statically import something you’d configured to split.

**Comment by fregante** [comment]
That'd be good if possible, but I don't expect all bundlers to be that configurable + automatic detection (via lint) is still always better than configuration.

**Comment by ljharb** [comment]
Are there any other use cases besides code-splitting where someone might want to force, or prohibit, a specific importing method?

**Comment by fregante** [comment]
I think generally any instance where you prefer delaying the loading/execution of a module, whether it's a module that is heavy, a polyfill that changes the environment, etc. 

I think any situation where `import()` would be useful over its static counterpart.

**Comment by ljharb** [comment]
Personally I'm thinking more of the reverse; other than code-splitting, I can't think of any time I'd prefer _dynamic_ import.

In particular, polyfills must *always*, without exception, run synchronously before all other code; otherwise they're not doing their job properly.

**Comment by fregante** [comment]
> In particular, polyfills must _always_, without exception, run synchronously before all other code; otherwise they're not doing their job properly.

Yes, but no. You don't _always_ want to load polyfills. So this would be reasonable:

```js
if (polyfillneeded) {
	await import('polyfill')
}

drawOwl()
```

**Comment by fregante** [comment]
> I can't think of any time I'd prefer _dynamic_ import.

Here, the second paragraph: https://github.com/tc39/proposal-dynamic-import#motivation-and-use-cases

**Comment by ljharb** [comment]
Yes, you definitely do - a) the polyfill itself is the only thing that should decide if it's needed, and b) the polyfill *must* load before ANY code has the chance to cache the original, unpolyfilled things.

**Comment by ljharb** [comment]
At any rate, I think there might be a case here for a rule that can be configured with an object that maps resolveable specifiers (they'll be run through the resolvers first) to an object of `{ require: true, import: true, dynamicImport: true }`.

**Comment by fregante** [comment]
I just remembered a situation where require/loading time time applied to Node as well:

- https://github.com/chalk/chalk/issues/180#issue-244858671

In short: The module was doing heavy operations on load, without any function calls, so people were thinking about [loading it only when needed.](https://github.com/chalk/chalk/pull/358/files#diff-88a99604221821c7258fe9c24fc58a4ef42242b97aeae97d6c438db3f62be6cdR208) Rare but possible.

**Comment by ljharb** [comment]
That's usually handled by conditional requires, not by switching to `import()`.

**Comment by fregante** [comment]
`import()` is exactly what's needed for those "conditional requires". There's no "conditional static import," that's why I'm saying it's exactly the same. If chalk was a ESM module back then they would have used exactly `import()` for this purpose.

**Comment by ljharb** [comment]
(To be clear, ESM and CJS are both first-class module systems in node for the duration, so there won't be any ESM-only functionality added here - https://github.com/import-js/eslint-plugin-import/issues/2326#issuecomment-992197221 is the thing I'm prepared to accept)

**Comment by fregante** [comment]
Found an exception to be aware of:

```js
import { type AutocompleteItem } from "autocompleter";

const {default: autocompleter} = import("autocompleter");
```

The first line should be ignored because it's just a type.

**Comment by ljharb** [comment]
O.o is that typescript? i thought TS had `import type` but not an inline `type` qualifier.

**Comment by fregante** [comment]
I think it's new. It's not required but it does make for clear examples.

**Comment by ljharb** [comment]
We should definitely support that then.

(That said, it seems much better to keep `import type`s and `import`s separate, to me)

**Comment by sushruth** [comment]
I have another use case for this - We always want some specific modules to be lazy-loaded in our application so they dont ever get accidentally added to our main bundle. For modules of this nature, I wish we could define a rule to always mandate dynamic imports for a particular module name pattern - example: `import(remoteModule/*)`.

I will try to follow this thread closely for this purpose.

**Comment by fregante** [comment]
As a recent fan of `no-restricted-syntax`, I just realized this can be done with that native eslint rule (although it's a bit verbose):

```js
"no-restricted-syntax": [
  "error",
  {
    selector: 'ImportDeclaration Literal[value="react"]',
    message: '"react" should only be imported via import("react")'
  },
]
```

<img width="733" alt="Screenshot 1" src="https://github.com/import-js/eslint-plugin-import/assets/1402241/3b6f4f34-a9e9-436c-ab75-306a11fe4d67">

To achieve the opposite ("Prefer import statement") replace `ImportDeclaration` with `ImportExpression`

This should be done with `no-restricted-imports` but they stale'd that request out:

- https://github.com/eslint/eslint/issues/13594

---

This request for a dedicated rule still stands though

```
"import/enforce-import-style": [
	"error",
	{
		static: ["react"],
		dynamic: ["my-little-heavy-dep"],
	}
]
```

**Comment by ljharb** [comment]
https://github.com/import-js/eslint-plugin-import/issues/2326#issuecomment-992197221 remains welcome for someone to send a PR.
